exporter/elasticsearchexporter/data_stream_router.go

```diff
@@ -0,0 +1,82 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package elasticsearchexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter"
+
+import (
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+func routeWithDefaults(defaultDSType, defaultDSDataset, defaultDSNamespace string) func(
+	pcommon.Map,
+	pcommon.Map,
+	pcommon.Map,
+	string,
+) string {
+	return func(
+		recordAttr pcommon.Map,
+		scopeAttr pcommon.Map,
+		resourceAttr pcommon.Map,
+		fIndex string,
+	) string {
+		// Order:
+		// 1. read data_stream.* from attributes
+		// 2. read elasticsearch.index.* from attributes
+		// 3. use default hardcoded data_stream.*
+		dataset, datasetExists := getFromAttributes(dataStreamDataset, defaultDSDataset, recordAttr, scopeAttr, resourceAttr)
+		namespace, namespaceExists := getFromAttributes(dataStreamNamespace, defaultDSNamespace, recordAttr, scopeAttr, resourceAttr)
+		dataStreamMode := datasetExists || namespaceExists
+		if !dataStreamMode {
+			prefix, prefixExists := getFromAttributes(indexPrefix, "", resourceAttr, scopeAttr, recordAttr)
+			suffix, suffixExists := getFromAttributes(indexSuffix, "", resourceAttr, scopeAttr, recordAttr)
+			if prefixExists || suffixExists {
+				return fmt.Sprintf("%s%s%s", prefix, fIndex, suffix)
+			}
+		}
+		recordAttr.PutStr(dataStreamDataset, dataset)
+		recordAttr.PutStr(dataStreamNamespace, namespace)
+		recordAttr.PutStr(dataStreamType, defaultDSType)
```

Comment:
This adds the three attributes to each record, even when the attributes were already found (e.g. on the resource). Is this necessary? Can we only add an attribute on the record when it wasn't found (using the `datasetExists` etc. variables)?

Reply:
quoting from @felixbarny from #33755 
> We should ensure that all docs have a data_stream.<type|dataset|namespace> property at the top-level, that's consistent with the _index ("${data_stream.type}-${data_stream.dataset:generic}-${data_stream.namespace:default}" when data_stream mode is enabled.

It makes sense to me for any docs sending to a data stream to have `data_stream.*` fields.

> I'm worried that this can inflate the request size. Is this necessary?

If by "request size" you meant bulk request size, yes, this is intended and desired.

If you meant adding useless attributes to attribute map that can be determined at document encoding time, and adding attributes to attribute map is slow, I agree that can be optimized, but what we have in this PR is good enough for now.

Reply:
Oh, I think my mind is running in circles. In fact, adding the attribute on the record in OTLP doesn't inflate the request size, because the document that is sent to Elasticsearch has the same size, whether the attribute is on the resource level or record level.

Let's just say it's too hot today to do code reviews. :sweat_smile:

Reply:
I think that ideally, we should not set the `data_stream.*` fields as attributes and instead just add them to the resulting ES document. It doesn't really matter for the ECS mode as every attribute is set at the top level, so we don't need to do these changes now.

> This adds the three attributes to each record, even when the attributes were already found (e.g. on the resource). I'm worried that this can inflate the request size.

For the ECS mode, there's no differentiation between different kinds of attributes as everything is added to the root of the doc, so there's no doc size impact.

A log record like this:

```yaml
body: "Hello world"
resource:
  attributes:
    data_stream.type: ignored # logs is enforced
    data_stream.dataset: foo-bar # will be sanitized, similar to https://github.com/elastic/elasticsearch/blob/f14c4a007e612890e42dc8c0e04cfdff65535808/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RerouteProcessor.java#L203-L206
```

should look like this in ECS mode:

```yaml
data_stream:
  type: logs
  dataset: foo_bar 
  namespace: default
message: "Hello world"
```

and should look like this in otel mode:

```yaml
data_stream:
  type: logs
  dataset: foo_bar
  namespace: default
body: "Hello world"
resource:
  attributes:
    data_stream.type: ignored
    data_stream.dataset: foo_bar
```

it should not look like this:

```yaml
data_stream:
  type: logs
  dataset: foo
  namespace: default
body: "Hello world"
resource:
  attributes:
    data_stream.type: ignored
    data_stream.dataset: foo_bar
    data_stream.namespace: default
attributes:
  data_stream.type: ignored
  data_stream.dataset: foo_bar
  data_stream.namespace: default
```

Reply:
thanks for pointing out that it will affect non-ECS modes. Do you think this is a blocker?

Reply:
No, definitely not a blocker. But let's do a follow up on that soon, as we're planning to work on the ECS mode imminently.

Reply:
@felixbarny do we need `data_stream` namespace in the semconv? Should I introduce it there?

Reply:
Maybe the doc should actually look like this in OTel mode:

```
data_stream:
  type: logs
  dataset: foo_bar
  namespace: default
body: "Hello world"
```

This would mean that we don't store `data_stream.*` values in the `*attribute.*` namespaces but always at the top-level.